### PR TITLE
Cluster Manager: Add support for managed VTEP mode

### PR DESCRIPTION
**Pull request by tssurya**
## 📑 Description

Adds support for managed VTEP mode from cluster manager side

1. an allocator logic is added
2. mutability logic is added (there is slight assymmetry in logic in how we do the mutation)
    1. Managed → Unmanaged: The IPs in the annotation were virtual allocations from the CM's pool. They don't correspond to any actual interface on the node. In unmanaged mode, the node-side controller needs to discover a real interface IP within the CIDRs. That discovered IP will almost certainly be different from what the CM allocated, so there's nothing meaningful to "preserve" — the managed IP was never on a real interface.
    2. Unmanaged → Managed: The IPs in the annotation came from real interfaces on the host. The CM can read them, mark them as allocated in the pool, and keep using them. The IPs are valid within the CIDRs and already "exist" — preserving them avoids unnecessary churn.
    3. adding new cidrs and expansion of mask is handled for managed mode
4. unit and integration tests added, e2e's will follow in the node side PR

## ✅ Checks

- [x] My code requires changes to the documentation
- [x] if so, I have updated the documentation as required
- [x] My code requires tests
- [x] if so, I have added and/or updated the tests as required
- [x] All the tests have passed in the CI 

## Summary by CodeRabbit

* **New Features**
  * Managed-mode VTEP IP allocation implemented end-to-end with per-VTEP allocators, initial sync/recovery, and managed→unmanaged cleanup.
  * Ability to widen CIDR ranges while preserving prior allocations and allocation order.
  * Subnet handling supports single-IP host subnets (/32 and /128).
  * Safe helpers for concurrent updates of node VTEP annotations.

* **Tests**
  * Large test expansion: allocator behavior, range replacement, single-IP allocations, managed-mode scenarios, integration coverage, and many edge cases.

* **Documentation**
  * EVPN docs: Managed mode guidance and mode-switching procedures.

### CODE REVIEW

File: go-controller/pkg/clustermanager/vtep/allocator.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+func (a *vtepIPAllocator) markAllocatedForNode(nodeName string, ips []net.IP) error {
+	ipNets := make([]*net.IPNet, 0, len(ips))
+	for _, ip := range ips {
+		mask := net.CIDRMask(ipv4HostSubnetLen, ipv4HostSubnetLen)
+		if utilnet.IsIPv6(ip) {
+			mask = net.CIDRMask(ipv6HostSubnetLen, ipv6HostSubnetLen)
+		}
+		ipNets = append(ipNets, &net.IPNet{IP: ip, Mask: mask})
+	}
+	return a.allocator.MarkAllocatedNetworks(nodeName, ipNets...)
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/allocator.go:108]: should you add vtep name in the error msg for better debuggability ?

File: go-controller/pkg/clustermanager/vtep/allocator_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+		v4count, v6count := a.allocator.RangeCount()
+		gomega.Expect(v4count).To(gomega.Equal(uint64(1)))
+		gomega.Expect(v6count).To(gomega.Equal(uint64(1)))
+
+		ips, err := a.allocateForNode("node-1")
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		gomega.Expect(ips).To(gomega.HaveLen(2))
+		gomega.Expect(ips[0].Mask).To(gomega.Equal(net.CIDRMask(32, 32)))
+		gomega.Expect(ips[1].Mask).To(gomega.Equal(net.CIDRMask(128, 128)))
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/allocator_test.go:45]: nit: can you invoke `a.allocateForNode("node-1")` again to check itempotent works ?

File: go-controller/pkg/clustermanager/vtep/allocator.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		return fmt.Errorf("failed to add CIDR %q to allocator: %w", cidr, err)
+	}
+	return nil
+}
+
+// allocateForNode allocates the next available IP per family for the given
+// node. The underlying allocator is idempotent: if the node already has an
+// allocation for a family it is returned as-is. Returns one /32 or /128
+// net.IPNet per configured family (IPv4 first, then IPv6).
+func (a *vtepIPAllocator) allocateForNode(nodeName string) ([]*net.IPNet, error) {
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/allocator.go:91]: i hope allocated IP from a subnet never changes until node is removed the cluster.

File: go-controller/pkg/clustermanager/vtep/allocator.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		hostLen = ipv6HostSubnetLen
+	}
+	if err := a.allocator.AddNetworkRange(ipNet, hostLen); err != nil {
+		return fmt.Errorf("failed to add CIDR %q to allocator: %w", cidr, err)
+	}
+	return nil
+}
+
+// allocateForNode allocates the next available IP per family for the given
+// node. The underlying allocator is idempotent: if the node already has an
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/allocator.go:88]: so it also implies allocated IP from a subnet never changes until node is removed the cluster.

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 5]
```diff
@@ -452,20 +528,131 @@ func (c *Controller) validateNodeVTEPIPs(vtep *vtepv1.VTEP) error {
...
+	//
+	// On retry we re-read from the informer cache (nodeLister), not the API
+	// server. If the informer hasn't caught up yet, the stale resourceVersion
+	// triggers another 409. DefaultBackoff provides 5 retries with
+	// exponential delays (10ms, 20ms, 40ms, …) which gives the informer
+	// time to sync. This is the same approach used by
+	// NodeAllocator.updateNodeNetworkAnnotationsWithRetry for the
+	// k8s.ovn.org/node-subnets annotation that has the same shared-key
+	// pattern across multiple network controllers.
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go]: as per commit message, we need to update node controller with UpdateNodeStatus + RetryOnConflict:
so can this move to util (vtep_node_annotations.go) with reusable functions ?

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 6]
```diff
@@ -48,11 +50,20 @@ type Controller struct {
...
 	cudnController controllerutil.Controller
 	nodeController controllerutil.Controller
 	eventRecorder  record.EventRecorder
 
+	// allocators holds per-VTEP IP allocators for managed-mode VTEPs.
+	// Keyed by VTEP name. Created on first reconcile, updated on CIDR
+	// changes, removed on VTEP deletion. Currently only accessed from
+	// vtepController (Threadiness=1), but protected by a mutex for
+	// safety if threadiness changes or cross-controller access is added.
+	allocatorsMu sync.Mutex
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go:63]: nit: should we more conservative about this by guarding `allocators` with a mutex ?  just adding a doc text not sufficient ?

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 7]
```diff
@@ -138,16 +138,80 @@ func NewController(
...
+			nodeVTEPs, err := util.ParseNodeVTEPs(node)
+			if err != nil {
+				if !util.IsAnnotationNotSetError(err) {
+					klog.Warningf("VTEP %s: failed to parse annotation on node %s, skipping: %v", vtep.Name, node.Name, err)
+				}
+				continue
+			}
+			entry, ok := nodeVTEPs[vtep.Name]
+			if !ok || len(entry.IPs) == 0 {
+				klog.Warningf("VTEP %s: no existing allocation found on node %s, skipping", vtep.Name, node.Name)
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go]: is this the case a new node joined to the cluster when vtep controller is down ? 
should it allocate new vtep ip here for the node ?

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 8]
```diff
@@ -195,19 +243,54 @@ func (c *Controller) reconcileVTEP(key string) error {
...
+		_, wasManaged := c.allocators[vtep.Name]
+		if wasManaged {
+			delete(c.allocators, vtep.Name)
+		}
+		c.allocatorsMu.Unlock()
+		if wasManaged {
+			klog.Infof("VTEP %s switched from Managed to Unmanaged, cleaning up annotations", vtep.Name)
+			if err := c.cleanupManagedVTEPAnnotations(vtep.Name); err != nil {
+				return fmt.Errorf("failed to clean up managed VTEP annotations for %s: %w", vtep.Name, err)
+			}
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/vtep/controller.go:293]: good catch !

File: go-controller/pkg/clustermanager/vtep/controller_test.go [hunk 9]
```diff
@@ -420,6 +420,317 @@ var _ = ginkgo.Describe("VTEP Controller", func() {
...
+			node4 := newNodeWithVTEPAnnotation("node-4", nil)
+			vtep := newVTEP("widen-vtep", vtepv1.VTEPModeManaged, "10.0.1.0/30")
+			start(vtep, node1, node2, node3, node4)
+
+			gomega.Eventually(func() (*metav1.Condition, error) {
+				return getVTEPCondition(fakeVTEP, "widen-vtep", conditionTypeAccepted)
+			}).WithTimeout(5 * time.Second).Should(gomega.HaveField("Status", metav1.ConditionTrue))
+
+			// Collect existing IPs before widening.
+			getNodeIP := func(nodeName string) string {
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller_test.go]: nit: getNodeVTEPIP

File: go-controller/pkg/clustermanager/vtep/controller_test.go [hunk 10]
```diff
@@ -420,6 +420,317 @@ var _ = ginkgo.Describe("VTEP Controller", func() {
...
+				}
+				return util.ParseNodeVTEPs(n)
+			}).WithTimeout(5 * time.Second).Should(gomega.HaveKey("widen-vtep"))
+
+			// Existing nodes must keep their original IPs.
+			gomega.Expect(getNodeIP("node-1")).To(gomega.Equal(ip1))
+			gomega.Expect(getNodeIP("node-2")).To(gomega.Equal(ip2))
+		})
+
+		ginkgo.It("preserves existing node IPs when switching from Unmanaged to Managed", func() {
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller_test.go:552]: for `Unmanaged` -> `Managed` switch, so here it may have CIDRs pointing to machine network subnet. how do we avoid IP(s) that are already assigned nodes for the primary network interface ?

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 11]
```diff
@@ -593,12 +599,29 @@ func (c *Controller) validateNodeVTEPIPs(vtep *vtepv1.VTEP) error {
...
+//  1. getOrUpdateAllocator: get the existing allocator, create one if this is
+//     the first managed reconcile (e.g. Unmanaged->Managed switch), or update
+//     it if the CIDRs changed. On first creation, existing node annotations
+//     are immediately marked so in-range IPs are preserved within this same
+//     reconcile.
+//  2. allocateAndAnnotateNode for every node: for nodes whose IPs were
+//     successfully marked above, allocateForNode is idempotent and returns the
+//     same IP. For nodes with stale/out-of-range IPs (mark was skipped), a
+//     fresh IP is allocated and the annotation is overwritten -- all in this
+//     same reconcile, not deferred.
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go:667]: so this is called even for annotation update from node side as well (in case node side update happened at the end after cm update for unamanged to manager case) ?   so it again updates the annotation with allocated ip ?

File: go-controller/pkg/clustermanager/clustermanager_test.go [hunk 12]
```diff
@@ -1649,6 +1656,314 @@ var _ = ginkgo.Describe("Cluster Manager", func() {
...
+							ClusterUserDefinedNetworkSelector: &apitypes.ClusterUserDefinedNetworkSelector{
+								NetworkSelector: metav1.LabelSelector{
+									MatchLabels: map[string]string{"evpn": "true"},
+								},
+							},
+						}},
+					},
+				}
+
+				// The RA is intentionally NOT pre-seeded. We start with CUDN+VTEP+FRR
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/clustermanager_test.go]: for my understanding: is it a requirement for end user as well ? should RA created before CUDN+VTEP+FRR ?

**Comment by tssurya** [comment]
thanks @pperiyasamy for the reviews! I'll work on these tomorrow

File: go-controller/pkg/clustermanager/vtep/allocator.go (hunk 4 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/allocator.go:88]: yup that's right, the vtep ip allocated for a node never changes until node is deleted from the cluster - to avoid causing loss of connectivity

File: go-controller/pkg/clustermanager/vtep/allocator.go (hunk 3 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/allocator.go:91]: that's right! ip allocated from a subnet for a node can never change until the node is removed from the cluster - ill make sure to add a unit test for this.

File: go-controller/pkg/clustermanager/vtep/allocator_test.go (hunk 2 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/allocator_test.go:45]: yea!! good point so that should also fix : https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6353/changes/dbdccae2e7af112f0cb734806b3d8f51a9d6bff3#r3329955722 this doubt

Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/allocator_test.go:45]: done.

File: go-controller/pkg/clustermanager/vtep/allocator.go (hunk 1 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/allocator.go:108]: yea will add it in the controller commit since at allocator level there is no notion of vtep..

File: go-controller/pkg/clustermanager/vtep/allocator.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+func (a *vtepIPAllocator) markAllocatedForNode(nodeName string, ips []net.IP) error {
+	ipNets := make([]*net.IPNet, 0, len(ips))
+	for _, ip := range ips {
+		mask := net.CIDRMask(ipv4HostSubnetLen, ipv4HostSubnetLen)
+		if utilnet.IsIPv6(ip) {
+			mask = net.CIDRMask(ipv6HostSubnetLen, ipv6HostSubnetLen)
+		}
+		ipNets = append(ipNets, &net.IPNet{IP: ip, Mask: mask})
+	}
+	return a.allocator.MarkAllocatedNetworks(nodeName, ipNets...)
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/allocator.go:108]: @coderabbitai: how can a node carry one valid and one invalid IP? is that scenario possible given syncNode does it in batch?
if we had one stale and one not - that's only possible during shrinking of cidr right? which is protected by CEL as a not possible scenario?

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 6 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:63]: hmm `allocatorsMu` is the mutex for `allocators` - maybe some confusion?
my comment also says:
"
Currently only accessed from
	// vtepController (Threadiness=1), but protected by a mutex for
	// safety if threadiness changes or cross-controller access is added.
"
so allocatorsMu protects full allocators

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 14]
```diff
@@ -246,11 +329,19 @@ func (c *Controller) handleVTEPDeletion(vtep *vtepv1.VTEP) error {
...
 	}
 
+	if vtep.Spec.Mode == vtepv1.VTEPModeManaged {
+		if err := c.cleanupManagedVTEPAnnotations(vtep.Name); err != nil {
+			return fmt.Errorf("failed to clean up managed VTEP annotations for %s: %w", vtep.Name, err)
+		}
+		c.allocatorsMu.Lock()
+		delete(c.allocators, vtep.Name)
+		c.allocatorsMu.Unlock()
+	}
```
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go]: good point, there is a narrow possible race here.

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 5 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go]: yes good point, let me do that - so that way in your PR you can change that bit on the node side

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 7 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go]: yes its the case when node joined when controller is down - but in this case since controller itself will once again do all add events - we don't need to take care of that here - so that should happen within the main controller right?
our goal with initial sync is to only deal with deletions and stale-ness

**Comment by tssurya** [comment]
> thanks @pperiyasamy for the reviews! I'll work on these tomorrow

sorry for the delay, specially that "tomorrow" became one sprint :/ 
Have to address one more comment: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6353#discussion_r3224702190 but I think I addressed the rest of them

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 11 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:667]: allocateForNode is idempotent so returns same IP and we have a guard ipsEqual that returns early - so no, in case annotations already exist and next reconcile is using same IP then no API annotation updates are done again

File: go-controller/pkg/clustermanager/vtep/controller_test.go (hunk 10 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller_test.go:552]: ah good question I think in the OKEP, we had some design discussion around this.. I think I do need to do node cidr overlap for managed case

File: go-controller/pkg/clustermanager/clustermanager_test.go (hunk 12 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/clustermanager_test.go]: no its not a requirement for end users - just the test gimmicks - for users whether they create the RA before or after doesn't matter, ill update the comment to make that clear (its just to avoid the state going to accepted=false for ra or pending config till NAD is created in the tests, in real prod this is not a concern)

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:293]: @coderabbitai: check the latest push and see if this is resolved

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go]: @coderabbitai: check the latest push and see if this is resolved

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 15]
```diff
@@ -452,20 +563,190 @@ func (c *Controller) validateNodeVTEPIPs(vtep *vtepv1.VTEP) error {
...
+	// VTEPs. Skip the write if the annotation already matches so we don't
+	// produce an infinite reconcile loop.
+	vteps, err := util.ParseNodeVTEPs(node)
+	if err == nil {
+		if existing, ok := vteps[vtepName]; ok && ipsEqual(existing.IPs, ips) {
+			return nil
+		}
+	}
+
+	return util.SetNodeVTEPEntry(nodeName, vtepName, ips, c.nodeLister.Get, c.kube.UpdateNodeStatus)
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go:839]: the node controller would use SSA mechanism to update vtep annotation `c.kube.SetAnnotationsOnNodeWithFieldManager` which helps to identify to ignore events triggered by its own update ? so here CM uses `c.kube.UpdateNodeStatus` which may not be SSA, but still  i guess it won't be a problem, right?

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 16]
```diff
@@ -529,28 +810,58 @@ func cudnNeedsUpdate(oldObj, newObj *udnv1.ClusterUserDefinedNetwork) bool {
...
+}
+
+// nodeNeedsUpdate triggers VTEP reconciliation on node creates and when the
+// k8s.ovn.org/vteps annotation changes.
+//
+// Creates (oldObj==nil) are accepted because managed-mode VTEPs need the
+// cluster-manager to allocate IPs for newly joined nodes. Without this, a
+// node joining a running cluster would not get a VTEP IP until an unrelated
+// event (e.g. a VTEP spec change) happened to trigger a VTEP reconcile.
+// For unmanaged mode the extra reconcile from a node create is harmless:
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go:955]: nit: can we just ignore the event if it is triggered from node controller side ?

```
	if util.IsLastUpdatedByManager(vtepAnnotationFieldManager, newNode.ManagedFields) {
		return false
	}
```

Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:955]: @pperiyasamy vtep annotation doesn't have any owners.. we aren't using SSA right? - ignoring event based on last update can be dangerous as well - cause maybe it changed in API but there is lag in informer that is still processing the previous event?

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 15 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:839]: node controller shouldn't use SSA - It also needs to use updateRetryConflict - else its going to overwrite things - the field manager concept doesn't work since the same annotation is owned by both components actually at the same time - only difference is each "vtep" in that annotation has separate ownership

File: go-controller/pkg/clustermanager/vtep/controller_test.go (hunk 10 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller_test.go:552]: the last commit should fix this as discussed in https://redhat-internal.slack.com/archives/C09EXCYBEEM/p1780298820942339

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 15 above)
Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/vtep/controller.go:839]: yes, good point, the node controller shouldn't use SSA now for updating vtep annotation. will just use `updateRetryConflict`.
I think there would be a corner case where potential race might happen with vtep annotation update while moving from unmanaged to managed.
1. cm's vtep controller processes vtep update (unmaged->managed>, allocates ip and update the vtep annotation.
2. vtep node controller still processing previous vtep unmanaged object version and update vtep annotation.

Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:839]: @pperiyasamy yes I forsee many tiny races to be honest with this design - its not great :( but that ship has sailed, if things are able to become "Eventually consistent" let's aim for that? we can accept some down time and all but if because of the stupid race things never become eventually consistent and we are stuck in bad state where node controller deletes what CM sets like you said - the only way around is to have some sort of signal between the two components.
In this case afterwards as long as CM re-takes the control back from node and node eventually stops the fight we are good, but I guess we kinda need to test these bits thoroughly to see how bad the race is

also note unmanaged to managed - the race won't have functional impact because CM mostly when it takes over - will try to accommodate the the same IPs the node gave initially, see my docs around mode switch. so in this case the allocated ip will be same as what node had as long as CIDR didn't change

tricker situation is managed to unmanaged

Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/vtep/controller.go:839]: yes @tssurya so i just now see `allocateAndAnnotateNode` handles it for unmanaged to managed transition. The `reconcileVTEP` handles VTEP transitions from Managed to Unmanaged, so this is already proves to be "Eventually consistent".  right ?

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 6 above)
Comment (pperiyasamy) [reply go-controller/pkg/clustermanager/vtep/controller.go:63]: oh! right, `allocators` is also being consumed by clustermanager-vtep-node-controller, so mutex is necessary.

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 17]
```diff
@@ -246,11 +258,23 @@ func (c *Controller) handleVTEPDeletion(vtep *vtepv1.VTEP) error {
...
-		// we don't retry here since when a CUDN is deleted, this controller will be notified and will re-queue the VTEP
 		klog.Infof("VTEP %s is still referenced by CUDNs [%s], blocking deletion", vtep.Name, strings.Join(referencingCUDNs, ", "))
 		return nil
 	}
 
+	c.allocatorsMu.Lock()
+	_, wasManaged := c.allocators[vtep.Name]
+	c.allocatorsMu.Unlock()
+
+	if wasManaged || vtep.Spec.Mode == vtepv1.VTEPModeManaged {
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go:363]: nit: why checking it twice ?

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 18]
```diff
@@ -137,16 +137,80 @@ func NewController(
...
+			ips := make([]net.IP, 0, len(entry.IPs))
+			for _, ipStr := range entry.IPs {
+				ip := net.ParseIP(ipStr)
+				if ip == nil {
+					errs = append(errs, fmt.Errorf("VTEP %s: invalid IP %q in annotation on node %s", vtep.Name, ipStr, node.Name))
+					continue
+				}
+				ips = append(ips, ip)
+			}
+			if err := allocator.markAllocatedForNode(node.Name, ips); err != nil {
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go]: looks like we may need to handle unmanaged -> managed transition (when controller was done) case at this point , we shouldn't attempt mark those IPs, instead remove IPs from node vtep annotation.

File: go-controller/pkg/clustermanager/vtep/controller.go [hunk 19]
```diff
@@ -267,6 +242,48 @@ func (c *Controller) reconcileVTEP(key string) error {
...
+	// There is a transient window between CM cleanup and node-side write
+	// where the VTEP annotation entry is absent. During this window
+	// validateNodeVTEPIPs reports AllocationFailed. The system self-heals
+	// once the node writes its discovered IP.
+	//
+	// Race with node-side during the cleanup window: the CM removes the VTEP
+	// entry from each node's annotation using RetryOnConflict+UpdateNodeStatus,
+	// while ovnkube-node concurrently writes its discovered IP using the same
+	// pattern (TODO: node-side should also use RetryOnConflict+UpdateNodeStatus;
+	// currently it uses strategic merge patch). Both sides may 409-conflict
```
Comment (pperiyasamy) [inline go-controller/pkg/clustermanager/vtep/controller.go:277]: right, doing it here https://github.com/ovn-kubernetes/ovn-kubernetes/pull/6378/changes/644cd3ed7b54aa620d500c7503d236acc5fea7be#diff-cf65cf169fd91887dd7bcba1300735c8c617e48986f33a2c984ebb44e8b39134.

**Review by pperiyasamy — COMMENTED** [review]
some minor nits.

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 17 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:363]: its not that we are checking it twice - its that we are saying "was it managed mode before and now its unmanaged but we need to clean up" OR "is it managed mode now.."

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 18 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go]: that's a good point...
but note that when we move from unmanaged to managed - as said in docs  - we want to try to keep the same IPs if possible, if the range didn't change...
for say the range did change -
note that the reason I didn't fix it before was
I thought The caller (handleManagedMode) will
// allocate a fresh IP and overwrite the stale annotation for those nodes
// within the same reconcile.
as in comments, but removing it is safer than ignoring it

Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go]: fixed this

File: go-controller/pkg/clustermanager/vtep/controller.go (hunk 15 above)
Comment (tssurya) [reply go-controller/pkg/clustermanager/vtep/controller.go:839]: correct!

**Comment by tssurya** [comment]
last push just fixed conflicts

File: docs/features/bgp-integration/evpn.md [hunk 20]
```diff
@@ -103,8 +103,10 @@ spec:
 ```
 
 > [!NOTE]
-> Managed VTEP mode is not yet implemented. Currently only unmanaged mode is
-> supported.
+> In **managed mode**, ovnkube-cluster-manager allocates VTEP IPs from the
```
Comment (tssurya) [inline docs/features/bgp-integration/evpn.md]: NIT: Actually this nit should just totally be gone..

**Comment by tssurya** [comment]
https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/27342044045/job/80783065211?pr=6353
```
Summarizing 1 Failure:
  [FAIL] e2e ingress to host-networked pods traffic validation Validating ingress traffic to Host Networked pods with externalTrafficPolicy=local [It] Should be allowed to node local host-networked endpoints by nodeport services
```
seems unrelated ill keep an eye on this
https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/27342044045/job/80783065158?pr=6353
```
Summarizing 1 Failure:
  [FAIL] Load Balancer Service Tests with MetalLB [It] Should ensure connectivity works on an external service when mtu changes in intermediate node [Feature:Service]
  /home/runner/work/ovn-kubernetes/ovn-kubernetes/test/e2e/service.go:2797
```
thats the known metallb flake that @pperiyasamy is looking into I think
https://github.com/ovn-kubernetes/ovn-kubernetes/issues/6094

**Comment by tssurya** [comment]
@coderabbitai: please review, why did you skip this PR?
